Added Fastapi endpoint for reserialize_dags

shubham-pyc opened the pull request:
**Why**
https://github.com/apache/airflow/issues/47844


**What**
Added REST ENDPOINT to replicate CLI functionality of  `airflow dags reserialize`

Updated endpoint `/public/dags/manage/reserialize`

shubham-pyc committed:
Added Fastapi endpoint for reserialize_dags

shubham-pyc committed:
Merge branch 'main' into feature/dags_reserialize_api

uranusjr committed:
Rewrite db usage in asset decorator operator (#47896)

This leaves much room for optimization. We really should use a cache
across argument resolution, outlet_events, and inlet_events. Refactoring
can be done later though; this works for now.

kaxil committed:
AIP-72: Get `TriggerDagRunOperator` working with Task SDK (#47882)

closes https://github.com/apache/airflow/issues/47499

`TriggerDagRunOperator` requires direct DB access to trigger DAG runs, which is not feasible under AIP-72 / Task SDK. Several approaches were considered:

1. **New Airflow API provider**
   - Move `TriggerDagRunOperator` to a new provider built on top of the HTTP provider and Airflow API.
   - This would allow DAG authors to configure a dedicated Airflow connection.
   - However, the Public API currently requires API tokens and does not support username/password authentication, making setup cumbersome.

2. **Pass Task Token to API**
   - Use Task Token from the controller DAG to authenticate API calls.
   - This would allow triggering DAGs in the same deployment using the execution API.
   - However, Task-identity tokens have not been implemented yet.

3. **Handle via Task Execution API (Chosen Approach)**
   - Raise a `TriggerDagRunOperator` exception.
   - The Task Runner catches this and invokes a new endpoint in the Task Execution API to trigger the target DAG.

Since (2) is not yet available, this PR/commit implements (3) as a temporary solution.

bugraoz93 commented on airflow/api_fastapi/core_api/datamodels/dags.py:
```diff
@@ -113,6 +114,32 @@ class DAGPatchBody(StrictBaseModel):
+        all_bundles = list(manager.get_all_dag_bundles())
+        all_bundle_names = {b.name for b in all_bundles}
+
+        if value is not None:
```
```suggestion
        if value:
```

bugraoz93 commented on airflow/api_fastapi/core_api/datamodels/dags.py:
```diff
@@ -113,6 +114,32 @@ class DAGPatchBody(StrictBaseModel):
+    def validate_bundle_names(cls, value):
+        """Validate bundle names format and check for duplicates."""
+        manager = DagBundlesManager()
+        all_bundles = list(manager.get_all_dag_bundles())
```
I am not sure if we need to init and load all the DAG bundles for each request. Maybe we can have a method in it to get_all_dag_bundles_by_name()

bugraoz93 commented on airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -339,3 +342,68 @@ def delete_dag(
+
+
+@dags_router.post(
+    "/manage/reserialize",
```
```suggestion
    "/reserialize",
```
I think we don't need additional `manage`. It would be simpler

shubham-pyc replied:
In that case you won't be able to have a DAG with ID: reserialize

because we have an API router with /public/dags/{dag_id}

bugraoz93 commented on airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -339,3 +342,68 @@ def delete_dag(
+        )
+
+    except Exception as e:
+        session.rollback()
```
This could maybe go to finally?

insomnes committed:
Add ability to pull XCom when passing multiple map_indexes (#46728)

* Add ability to pull XCom when passing multiple map_indexes

* Add comment about LazyXComSequence and orignal description

* Apply review comments

* Update task_sdk/src/airflow/sdk/execution_time/task_runner.py

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

* Fix review comments

---------

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>
Co-authored-by: Amogh Desai <amoghrajesh1999@gmail.com>

bugraoz93 commented on airflow/api_fastapi/core_api/routes/public/dags.py:
```diff
@@ -339,3 +342,68 @@ def delete_dag(
+    except Exception as e:
+        session.rollback()
+        raise HTTPException(
+            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
```
I'm not sure about this broad status code. It should also be documented in openapi spec.

bugraoz93 reviewed (commented):
Thanks for the changes! Added suggestions

gopidesupavan committed:
Fix typos in task sdk (#47945)

jason810496 committed:
Fix k8s test: Ensure api-server healthy before any API calls (#47938)

Fix AttributeError

shubham-pyc committed:
Update airflow/api_fastapi/core_api/datamodels/dags.py

Co-authored-by: Bugra Ozturk <bugraoz93@users.noreply.github.com>

uranusjr committed:
Correctly select multiple columns with execute() (#47928)

Using scalars() silently drops the second object.

amoghrajesh committed:
Fix broken task mapping due to XCom refactor (#47944)

simonprydden committed:
Fix mypy boto3 appflow version (#47912)

* refactor: remove type casting in AppflowHook update_flow method

* chore: update mypy-boto3-appflow dependency to version 1.37.0

* chore: update mypy-boto3-appflow dependency to version 1.37.0

* chore: update mypy-boto3-appflow dependency to version 1.37.0

---------

Co-authored-by: Simon Prydden <mea99sd@aol.com>
Co-authored-by: Kalyan R <kalyan.ben10@live.com>

bbovenzi committed:
Add asset event extra to events list in asset details page (#47935)

bbovenzi committed:
Add next run to asset expression (#47932)

* Add next run assets data to Dag Schedule component

* Update popover bg colors

timonviola committed:
docs: enhance airflowLocalSettings value description (#47855)

dstandish committed:
Next run date should not be updated by backfills (#47922)

If we update next run after a backfill, for a dag with catchup enabled, the scheduler may create old runs that are before start date (in the period between the last backfill run and the start date of the dag).

uranusjr committed:
Delete asset relations when dag becomes inactive (#47903)

* Delete asset relations when dag becomes inactive

This syncs the relations when a dag is deleted in dag files. Makes the
"consuming_dags" and "producing_tasks" display correctly.

* Add test for auto asset ref removal

jedcunningham committed:
Fix code blocks in extra packages reference docs (#47960)

kaxil committed:
Fix bug in XComModel deserialization for value serialized in Task SDK (#47961)

Currently, the Task SDK uses `airflow.serialization` module to serialize XCom value during execution. This is because
we have agreed with the contract of the execution client (Python, Golang and other SDK clients) being responsible for serialization and de-serialization.

To do this, as part of https://github.com/apache/airflow/issues/45481, we skip XCom serialization from XComModel and instead directly store in the DB.

This works well when deserialization is done on Execution side itself. However, for some cases we deserialize it on the Scheduler or API server only. Some of those:
- Extra Operator Links
- Branching & Skipping logic in `NotPreviouslySkippedDep` where we pull XCom.

closes https://github.com/apache/airflow/issues/47907
closes https://github.com/apache/airflow/pull/47918

jedcunningham committed:
Fix typo in unit test docs (#47964)

uranusjr committed:
Set dag version on expanded tis (#47953)

The version is set to the unmapped ti's if available, latest otherwise.

uranusjr committed:
Only return active assets in API by default (#47962)

Similar to the list dags endpoint, the list assets endpoint now has a
query param 'only_active' that defaults to true (i.e. only show active
assets). You can set this to false for the endpoint to also show
inactive (deleted from code) assets.

kaxil committed:
Fix typo in ``InProcessExecutionAPI`` (#47966)

jedcunningham committed:
Move FAB session table creation to FAB provider (#47969)

We need to create the `session` table in the provider db manager, not in
the core db utils.

Co-authored-by: vincbeck <vincbeck@amazon.com>

jedcunningham reviewed (changes requested):
First off, thanks for opening a comprehensive first PR! It looks great.

Unfortunately, we can't add an endpoint that will reserialize all of the DAGs in the api-server process. If you look at our security model, we state that DAG files are not necessary for the scheduler and api-server (it says webserver, but that just hasn't been updated yet 😄), and most importantly, that user provided code should not be executed in those components.

https://github.com/apache/airflow/blob/main/docs/apache-airflow/security/security_model.rst#dag-files-not-needed-for-scheduler-and-webserver

If you look at how the [reparse a file feature](https://github.com/apache/airflow/blob/2500dcf20d2782d16da53ee857c0aab21bfdfbf2/airflow/api_fastapi/core_api/routes/public/dag_parsing.py#L47) works, you'll see it needs to communicate with the dag processor via the db in order to carry out reparsing of a specific file. We will need to follow a similar path to do a "reparse everything" feature.

I have a few ideas on how we could do this, but with feature freeze on Airflow 3 being tomorrow (really closer to the end of today 😄), unfortunately, this will have to wait for 3.1 so we have a bit of time to iterate on it.

ferruzzi committed:
Update example_s3 system test (#47974)

A Sensor was being skipped for a reason that is no longer true, so we may as well stop skipping it.

bugraoz93 committed:
AIP-81 - API Communication Mechanism (#45300)

jedcunningham committed:
Default to the FabAuthManager in the chart (#47976)

Core is going to move to the SimpleAuthManager by default, but the chart
should keep using the FabAuthManager as a default instead for the time
being. Users can always change this to another auth manager if they need
to.

Commit:
AIP-84 Move public api under /api/v2 (#47760)

yunchipang committed:
docs: fix numbered list (#47978)

shubhamraj-git committed:
Add Create Connection button and form on connection page (#47868)

* Move useParamStore to common

* Foundation work on connection add form

* Added Form

* Final checks

* Reviews

* Reviews

* Move Extra JSON into Accordion

---------

Co-authored-by: Jens Scheffler <95105677+jscheffl@users.noreply.github.com>

pykenny committed:
Update task lifecycle diagram for missing states (#46056)

* Updates task lifecycle diagram for missing states

Details:
 - Adds missing states to the diagram ("skipped" and
   "deferred")
 - Adds missing condition nodes to state transition
 - Adds "trigger" component to the diagram
 - Coverts multiple conditional branch in the graph into
   sequence (multi-staged) of binary conditional branches
 - Split states into three categories: "shared states",
   "states only for sensors", and "states only for deferrable
   tasks"
 - Uses different icons to differentiate task states and Airflow
   components

* Adds transitions for triggerer tasks

Details:
 - Adds conditions and transitions for tasks that only require
   triggerer to execute

* Make cycles look better

* [WIP] Initial attempt for generating diagram

Details:
 - Adds icon images to docs/diagrams
 - Adds 'diagram_task_lifecycle_diagram' script to generate
   lifecycle diagram with `diagrams`
 - Initial attempt to build up the graph, without label for
   condition branch edges, and some redundant edges

* Adjusts branch labels; Adds label to condition branches

Details:
 - Adjusts branch label location by padding newline and
   whitespace
 - Uses `Edge` constructor to add label for condition branch edges
 - Removes redundant edges and correct incorrect edge directions

* Remove original image; Update documentation

* Adds legend cluster

* File renaming

Details:
 - "diagram_task_lifecycle_diagram" > "diagram_task_lifecycle"

amolsr committed:
Xcoms object storage doc clean (#47924)

* doc: fix contribution doc ports (#47176)

* doc: clean up for xcoms docs

uranusjr committed:
Implement on_execute_callback in task sdk (#47989)

The callable is run *outside* of the timeout block. Failures calling
any callback functions are simply logged without failing the entire
task. This matches Airflow 2's behavior.

Submitting this first to get feedback. Other callbacks will be
implemented in later PRs if I'm on the right track.

potiuk committed:
Add lower-binding for microsoft azure/winrm providers (#47990)

For Python 3.12 removal of distutils causes older versions of
microsoft libraries to fail when trying to import it when lowest
bound versions of msal-extensions and pywirm are used. Adding
lower binding for msal-extensions and pywinrm should avoid that.

ashb committed:
Introduce API versioning into the Execution API (#47951)

Right now we have exactly one version, but using Cadwyn gives us the ability
to make changes to the Execution API almost "at will" and it will
transparently upgrade from older request versions. The tl;dr of it is "apply
migrations to your requests so that you only have to maintain the latest
version".

From it's docs[^1]:

> Cadwyn allows you to support a single version of your code while
> auto-generating the schemas and routes for older versions. You keep API
> versioning encapsulated in small and independent "version change" modules
> while your business logic stays simple and knows nothing about versioning.

This gives us the freedom to change the version of the API but continue to
support "all" old versions. This is a key aspect of proposal in AIP-72. We
don't yet have any new versions -- those will come later.

[^1]: https://docs.cadwyn.dev/

kacpermuda committed:
feat: Add helper for any provider version check (#47909)

Lee-W committed:
fix(asset): allow extra in Metadata to be optional (#47997)

close: #47779

Lee-W committed:
docs(newsfragments): migration rules format fix (#47991)

amoghrajesh committed:
Move Secrets Backend for Variables & Connections support to Task SDK (#47946)

* Move Secrets Backend for Variables & Connections support to Task SDK

* adding support for variables too

* introducing config for workers

* moving the internal utils outside the objects

* adding task sdk tests

* coagulating into _get(connection, variable)

* adding secrets tests and task sdk tests

* add back removed stuff

* allowing fall back onto servers backend

vincbeck committed:
Set simple auth manager as default (#47691)

Dong-yeong0 committed:
Fix apache#47555: PostgreSQL URI driver issue  (#47709)

* Fix apache#47555: PostgreSQL URI scheme issue

* Fix: allow configuring PostgreSQL connection driver

Added support for customizing the PostgreSQL connection driver (`postgres` vs `postgresql`)
via the `postgres_driver` option in the extra JSON field.

* Fix apache#47555: PostgreSQL URI scheme issue

* Fix: allow configuring PostgreSQL connection driver

Added support for customizing the PostgreSQL connection driver (`postgres` vs `postgresql`)
via the `postgres_driver` option in the extra JSON field.

* Add the documentation for the 'designator' parameter in the PostgreSQL extras section

* Remove the code that constructs the URI based on connection type

* Update get_uri method docstring

* Clarify that get_uri() does not return a SQLAlchemy URI

* Fix rst-backticks

* Remove designer description

* Fix format code with ruff

---------

Co-authored-by: LIU ZHE YOU <68415893+jason810496@users.noreply.github.com>

dstandish committed:
Replace DAGs with dags in docs (#47959)

Reading "DAGs" everywhere in the docs is ugly, unnatural, and unnecessary. We don't always need to make a, perhaps a bit pedantic, reference to the class name when talking about dags. And, as discussed on the dev list, there seemed to be a general agreement that reference to the "directed acyclic graph" is not particularly helpful. So, here, I change most instances of DAGs in the docs to dags.

kacpermuda committed:
fix: compat test test_provider_not_installed failing on main (#48012)

pierrejeambrun committed:
AIP-38 Improve handling of mapped task group (#48010)

Closes: #47784

kaxil committed:
Bump OL provider for Airflow 3.0.0b4 release (#48011)

Bump OL provider for Airflow 3.0.0b4 release

vincbeck committed:
Call `init` from auth managers only once (#47869)

ashb committed:
Improve the error message when a DAG tries to access the DB. (#48015)

Previously it just gave a rather opaque error:

> module 'airflow.settings' has no attribute 'engine'

This improves that so that hopefully the user knows what is going on and isn't
left wondering if there's a bug in Airflow.

bugraoz93 committed:
AIP-81 Move remote CLI code to airflowctl (#47981)

This will set us up to have a completely standalone package,
which is a good goal since airflow has so many deps.

---------

Co-authored-by: Jed Cunningham <jedcunningham@apache.org>

bbovenzi committed:
Fix check for asset expression in Schedule component (#48020)

kaxil committed:
Bump various providers in preparation for Airflow 3.0.0b4 (#48013)

* Bump various providers in preparation for Airflow 3.0.0b4

Bumped providers:
- Celery
- K8s
- common.compat
- common.io
- common.messaging
- common.sql
- fab
- standard

* fixup! Bump various providers in preparation for Airflow 3.0.0b4

* Apply suggestions from code review

* Update providers/cncf/kubernetes/docs/changelog.rst

Co-authored-by: Elad Kalif <45845474+eladkal@users.noreply.github.com>

---------

Co-authored-by: Elad Kalif <45845474+eladkal@users.noreply.github.com>

ashb committed:
Improve handling around stopping/killing individual dag file processors (#48019)

We've had occasional reports of the following stack trace from the Dag
processor:

```
    self.processor.terminate()
  File "/opt/airflow/airflow/dag_processing/manager.py", line 1020, in terminate
    processor.kill(signal.SIGTERM, escalation_delay=5.0)
  File "/opt/airflow/task-sdk/src/airflow/sdk/execution_time/supervisor.py", line 544, in kill
    exit_code := self._service_subprocess(max_wait_time=end - now, raise_on_timeout=False)
  File "/opt/airflow/task-sdk/src/airflow/sdk/execution_time/supervisor.py", line 601, in _service_subprocess
    need_more = socket_handler(key.fileobj)
  File "/opt/airflow/task-sdk/src/airflow/sdk/execution_time/supervisor.py", line 960, in cb
    gen.send(line)
StopIteration
```

This should now correctly handle this case.

And another thing noticed in testing this I noticed that the duration for
in-process parsing processes was incorrect.

ahmadtfarhan committed:
Rename Breeze variables.env file to environment_variables.env (#47998)

o-nikolas committed:
Don't expect default connections to be present (#47968)

Airflow does not create default connections by default and this operator
was expecting them to be present and throwing exceptions. We should only
expect the connection to be present if the user explicitly passed that
connection name to us (rather than receiving it as a default value)

brouberol committed:
Reflect new base_container_status_polling_interval in the TaskDecoratorCollection type (#47833)

This is a follow up to https://github.com/apache/airflow/pull/47774,
addind the type of the new `base_container_status_polling_interval` argument
to the `TaskDecoratorCollection` type, to update the type autocompletion
when using the `@kubernetes` decorator.

Signed-off-by: Balthazar Rouberol <brouberol@wikimedia.org>

kaxil committed:
Export `chain`, `chain_linear` & `cross_downstream` (#48033)

Added `chain`, `chain_linear`, and `cross_downstream` to `__all__` in
`BaseOperator`, ensuring they are properly exported for use.

darkag committed:
fix inconsistence between the get_conn and run in `WinRMHook` (#47229)

* Change get_conn behaviour in order to make run callable multiple times

* fix tests

bugraoz93 committed:
Resolve conflict with uv sync and update httpx, change name of the CLI name to airflowctl, remove print (#48041)

bugraoz93 committed:
Remove not needed mypy check, it is already excluded now (#48039)

Prab-27 committed:
Move "create db from orm" to be a public method in db manager interface (#48000)

* moving create_db_orm to be a public method in db manager

* move create_db_from_orm to be a public method in fab provider

* Update providers/fab/src/airflow/providers/fab/auth_manager/models/db.py

---------

Co-authored-by: Vincent <97131062+vincbeck@users.noreply.github.com>

joeycumines-scw committed:
Refactor container image to persist msodbcsql18 eula acceptance (#47982)

* Update Dockerfile to persist msodbcsql18 EULA acceptance

* Update the other scripts

pierrejeambrun committed:
AIP-38 Fix SimpleAuthManager prefix (#48025)

Lee-W committed:
feat(api_fastapi): include asset id in asset nodes when calling "/ui/dependencies" and "/ui/structure/structure_data" (#47381)

* feat(serialization): change dependency_id in DagDependency with asset type to be asset_id

* feat(dag_dependency): add label

* test(serialization): fix test_derived_dag_deps_sensor

* test(serialization): fix test_dag_deps_assets_with_duplicate_asset

* test: fix test_derived_dag_deps_operator

* test(task_sdk): fix task_sdk tests due to _resolve_assets renaming

* test(api_fastapi/ui): fix test cases

* test(serialized_dag): fix test_order_of_deps_is_consistent

* feat(dot_renderer): support label

* refactor: refactor retreive_asset_models as resolve_assets_as_dag_dependencies to retreive needed info only

* feat: remove db access in serialized object

* feat(dag_dependency): improve dependency_type typing and split asset-ref to asset-name-ref and asset-uri-ref

* feat(serialized_dag): resolve dag dependency to include asset id

* refactor(serialized_dag): extract dag dependency resolving logic as a class

* test(dag_serialization): fix dag_serialization tests

* test(api_fastapi): update test cases with asset_id changes

* test(api_fastapi): fix test case with dynamic asset id

* fix(serialized_dag): fix postgrest dag dep handling

* Fix asset graph in UI

* test(api_fastapi): fix structure endpoint test case

* fix(api_fastapi): fix /ui/structure

---------

Co-authored-by: Brent Bovenzi <brent.bovenzi@gmail.com>

rawwar committed:
add kalyan in CODEOWNERS (#48045)

amoghrajesh committed:
Adding codeowners to task sdk (#48048)

kacpermuda committed:
fix: Re-add configuring orm for OpenLineage's listener on scheduler (#48049)

amoghrajesh committed:
Adding newsfragment for workers secrets backend (#48047)

potiuk committed:
Move airflow sources to airflow-core package (#47798)

This is continuation of the separation of the Airflow codebase into
separate distributions. This one splits airflow into two of them:

* apache-airflow - becomes an empty, meta no-code distribution that
  only has dependencies to apache-airflow-core and task-sdk
  distributions and it has preinstalled provider distributions
  added in standard "wheel" distribution. All "extras" lead
  either to "apache-airflow-core" extras or to providers - the
  dependencies and optional dependencies are calculated differently
  depending on "editable" or "standard" mode - in editable mode,
  just provider dependencies are installed for preinstalled providers
  in standard mode - those preinstalled providers are dependencies.

* the apache-airflow-core distribution contains all airflow core
  sources (previously in apache-airflow) and it has no provider
  extras. Thanks to that apache-airflow distribution does not
  have any dynamically calculated dependencies.

* the apache-airflow-core distribution hs "hatch_build_airflow_core.py"
  build hooks that add custom build target and implement custom
  cleanup in order to implement compiling assets as part of the build.

* During the move, the following changes were applied for consistency:

  * packages when used in context of distribution packages have been
    renamed to "distributions" - including all documentations and
    commands in breeze to void confusion with import packages
    (see
https://packaging.python.org/en/latest/discussions/distribution-package-vs-import-package/)

  * all tests in `airflow-core` follow now the same convention
    where tests are in `unit`, `system` and `integration` package.
    no extra package has been as second level, because all the
    provider tests have "<PROVIDER>" there, so we just have to avoid
    naming airflow unit."<PROVIDER>" with the same name as provider.

  * all tooling in CI/DEV have been updated to follow the new
    structure. We should always build to packages now when we
    are building them using `breeze`.

Commit:
Setting Airflow context Environment variables for operators (#47644)

Co-authored-by: Sneha Prabhu <snehaprabhu@Snehas-MacBook-Pro.local>
Co-authored-by: Amogh Desai <amoghrajesh1999@gmail.com>

jedcunningham committed:
Use SimpleAuthManager for standalone (#48036)

This fixes the standalone command.

At the same time, it switches standalone to use SimpleAuthManager.
Similar to before, if it's the first time you run, the username and
password will be printed out - now just at the very top of the output
(which honestly is a bit easier to see before a ton of lines start
 flying by). However, it will only output the password once. On
subsequent starts, it will print the file that contains the file so you
can go grab the password yourself if you need it. This was done since
there isn't an easy way to both use the same password with standalone and
api-server directly and make sure we aren't outputting a password you've
populated yourself.

bbovenzi committed:
Fix api server command (#48062)

danielvdende committed:
Fix execution api server url in docker-compose (#48061)

This was using AIRFLOW__WORKERS instead of AIRFLOW__CORE, which didn't work

dominic-lcw committed:
Add download button to export flowchart (#47890)

* Add download button to export flowchart

* refactored DownloadButton and updated styles

* Update toaster import path

pierrejeambrun reviewed (commented):
Yes we need to use `DagPriorityParsingRequest` to trigger the reparsing of appropriate files. There is no `DagModel.name` I guess you mean `DagModel.bundle_name` ?

Also people can use no bundle I think. (bundle_name is None). Therefore if this endpoints is only accessible providing the `bundles_to_process` such setup will not be able to use this feature. Maybe using the dagfolder. (there can be multiple dag folders), no args at all and we reparse 'everything'.

amoghrajesh committed:
Adding codeowners to task execution API (#48053)

eladkal committed:
update code owners (#48068)

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

kaxil committed:
Move newsfragment about secret backends on workers to AIP-72 one (#48069)

So that all AIP-72 ones remain on the same place.

uranusjr committed:
Calculate retry eligibility before task runs (#47996)

The task needs this information to know whether it should run the retry
or failed callback if the task fails, so we calculate it eagerly and
pass it to the runner, instead of lazily only after the task has failed.

The information is used by the task runner to decide what state to set
the ti to when an error happens. If the task is eligible for retry, the
next state is set to UP_FOR_RETRY; otherwise it's FAILED. This is then
handled accordingly in the API.

The FAIL_WITHOUT_RETRY state has been removed from the code base since
it is no longer needed. The state was introduced in 1283cc347f26678e7504
to tell the API server to not check retry eligibility on certain failure
cases, but now that eligibility is always checked and used in the task
runner, the API server no longer needs this information; a request to
update to FAILED always set the TI to FAILED.

ashb committed:
Config improvments to using JWKS with the JWT code (#48054)

Previously it was hard-coded to only use the thumbprint of the privkey as the
`kid`, which is a sensible default but not the only way JWKS documents work,
and it has to match between the generator and validator.

Additionally the JWTValidator had no way of specifying the algorithm(s) to use
when using a JWKS document and it would always fail. This now respects the
same config option that the JWTGenerator looked at.

ellisms committed:
Updating EC2 Operators and Sensors with AWS Base classes (#47931)

* Update ec2 opereator with AwsBaseOperator

* Update ec2 sensor with AWS base class

* Use AWS Base classes in EC2 operators and sensors

* Fixing static check issues

* PR feedback changes

---------

Co-authored-by: mse139 <mse139@users.noereply.github.com>

Narendra-Neerukonda committed:
Fixes dag version interpolation in UI (#48072)

tirkarthi committed:
Handle none logical date and show dependency check for the relevant state (#48067)

* Fix table header alignment and show blocking dependencies on relevant task state.

* Skip task start date dependency check when task instance logical date is None.

amoghrajesh committed:
Fixing upgrade checks on main (#48073)

kaxil committed:
Support `@task.bash` with Task SDK (#48060)

o-nikolas committed:
initdb has been gone for a while. Log the correct command (#48038)

In certain cases we detect that a DB initialization/migration is
needed and we log to the user the `initdb` command which no longer
exists (since 2.7.0). We should instead advise them to run migrate.

bbovenzi committed:
Add asset id to link between nodes in asset graph (#48022)

Fix test changes

Co-authored-by: Wei Lee <weilee.rx@gmail.com>

amoghrajesh committed:
Adding .vite from auth/managers to gitignore (#48078)

gopidesupavan committed:
Fix python operators errors when initialising plugins in virtualenv jinja script (#48035)

* Fix python operators errors when initializing plugins

* Update reason for setting the PYTHON_OPERATORS_VIRTUAL_ENV_MODE environment variable in jinja

vincbeck committed:
Add `vincbeck` as owner of fab provider, common messaging provider and auth manager (#48080)

AutomationDev85 committed:
Add ti.running metric export (#47773)

Co-authored-by: AutomationDev85 <AutomationDev85>

dheerajturaga committed:
Unquote executor airflow.cfg variable to fix #48042 (#48064)

When setting the executor in the Airflow configuration file (airflow.cfg), using executor='LocalExecutor,CeleryExecutor' causes a failure, while using executor = LocalExecutor,CeleryExecutor works correctly.

closes: #48042

amoghrajesh committed:
Fixing new update installer check errors in CI (#48081)

* Fixing new upgrade checks issues in CI

* upgrade fixes

* undoing some random update

pierrejeambrun committed:
Auth Manager should raise an invalid token error when the payload cann… (#48056)

* Auth Manager should raise an invida token error when the payload cannot be deserialize

* Fix test caching issue

jedcunningham committed:
Swap contributing links to absolute links (#48085)

Most of these work as a relative link, but they are fragile - if you get
the level of nesting wrong they break. Absolute is more resilient.

jroachgolf84 committed:
Changing triggerer config option `default_capacity` to `capacity` (#48032)

jedcunningham committed:
Bump cadwyn min version to 5.1.3 (#48091)

This fixes static checks on main, and ensures when we generate
consistent datamodels.

dependabot[bot] committed:
Bump trove-classifiers from 2025.3.13.13 to 2025.3.19.19 (#48023)

Bumps [trove-classifiers](https://github.com/pypa/trove-classifiers) from 2025.3.13.13 to 2025.3.19.19.
- [Release notes](https://github.com/pypa/trove-classifiers/releases)
- [Commits](https://github.com/pypa/trove-classifiers/compare/2025.3.13.13...2025.3.19.19)

---
updated-dependencies:
- dependency-name: trove-classifiers
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

potiuk committed:
Fix test directory structure for airflow-ctl distributio (#48093)

The "tests" folder is root folder for all the tests, it should not be
a Python package - similarly as in all other distributions, the "tests"
folder should be just root folder for all test python packages.

Also - we should not have some "generic" package name as the top
packages in "tests" folder - becuase that has high potential for
conflicting with other - similarly named packages, the convention
we use for all the packages is one of the two:

* unit,integration,system as top level (providers/airflow-core) in
  case the distribution has those three different type of tests

* "distribution_folder_name" (with underscore instead of dash)
  when only unit tests are expected to be in the module (for
  example in "task_sdk"

The `airflow-ctl` is rather unit-test-only distribution, so the
convention to avoid conflicts is `tests/airflow_ctl`. This
is what we use here.

bbovenzi committed:
Add dark/light mode to simple auth manager (#48065)

amoghrajesh committed:
Correcting AIRFLOW_ROOT_PATH to fix CI (#48077)

potiuk committed:
Make wtforms optional for Providers Manager (#48094)

amoghrajesh committed:
Fixing permission for aws-credentials in pushing docs (#48095)

jedcunningham committed:
Remove unused webserver configs (#48066)

This is my first pass through them - these are all unused.

jedcunningham committed:
Remove accidentally committed file (#48096)

amoghrajesh committed:
Do not allow semicolon in partition_clause (#48098)

niklasr22 committed:
make docker swarm service name customizable (#47957)

potiuk committed:
Remove breeze's top-level init file (#48101)

The `src` folder is a folder, not Python package and the
file __init__.py there is just wrong :)

potiuk committed:
Fix progress retrieval when parallel airflow-core tests are run (#48100)

mobuchowski committed:
serialize http transports contained in composite transport (#47444)

Signed-off-by: Maciej Obuchowski <maciej.obuchowski@datadoghq.com>

MaksYermak committed:
Add DNS endpoint support for GKE Hooks and Operators (#48075)

eladkal committed:
Update pre-commit setup docs to suggest uv (#48074)

eladkal committed:
Add clarification for sql injection in Airflow security model (#48016)

jedcunningham committed:
Smaller distribution files (#48089)

We are now including node_modules into our wheel/sdist files, which
makes them rather large. This brings them back down.

This was a regression with the core move.

jedcunningham committed:
Remove password extra - nothing uses those deps (#48027)
